### PR TITLE
Add Tests for ActionSequenceFinder

### PR DESCRIPTION
**Pull request by Entenwilli**
This PR adds tests for the ActionSequenceFinder using the BranchingOnlineShop, InternationalOnlineShop and TravelPlanner examples. 

The tests for each example are comprised of the following elements:
- analysis calculates results regarding dataflow in the model (e.g. the list of found sequences is not empty)
- correct amount of ActionSequences were found
- Types/Classes in each ActionSequence are correct
- content/entity name for elements in the ActionSequence are correct

### CODE REVIEW

File: tests/org.palladiosimulator.dataflow.confidentiality.analysis.tests/src/org/palladiosimulator/dataflow/confidentiality/analysis/DataFlowConfidentialityAnalysisTest.java [hunk 1]
```diff
@@ -2,58 +2,99 @@
...
+        assertEquals(allSequences.size(), 2);
 
-        assertSequenceElements(allSequences.get(0), CallingUserActionSequenceElement.class,
-                CallingSEFFActionSequenceElement.class, SEFFActionSequenceElement.class,
-                CallingSEFFActionSequenceElement.class, SEFFActionSequenceElement.class,
-                CallingUserActionSequenceElement.class, CallingUserActionSequenceElement.class,
-                CallingSEFFActionSequenceElement.class, CallingSEFFActionSequenceElement.class,
-                CallingUserActionSequenceElement.class);
+        assertSequenceElements(allSequences.get(0), 
+        		CallingUserActionSequenceElement.class, // UserAction-Call: ViewEntryLevelSystemCall
```
Comment (sebinside) [inline tests/org.palladiosimulator.dataflow.confidentiality.analysis.tests/src/org/palladiosimulator/dataflow/confidentiality/analysis/DataFlowConfidentialityAnalysisTest.java:73]: As comments have no semantics, these comments could become outdated quite easily. I would suggest either
a) Remove the comments (although I appreciate the effort :))
b) Create separate test cases that check for validity, e.g., of the names in named entities. This could build on the functions for checking the element content you already added.

Comment (sebinside) [reply tests/org.palladiosimulator.dataflow.confidentiality.analysis.tests/src/org/palladiosimulator/dataflow/confidentiality/analysis/DataFlowConfidentialityAnalysisTest.java:73]: Nevertheless, I would not focus on such detailed test cases for now. So maybe go with b) only if you have spare time besides finishing the analysis implementation :)

**Review by sebinside — CHANGES_REQUESTED** [review]

File: tests/org.palladiosimulator.dataflow.confidentiality.analysis.testmodels/models/InternationalOnlineShop/results.txt [hunk 2]
```diff
@@ -0,0 +1,11 @@
+-------------- GENERAL ---------------
```
[the remark this answers is not in the document]
Comment (Entenwilli) [reply tests/org.palladiosimulator.dataflow.confidentiality.analysis.testmodels/models/InternationalOnlineShop/results.txt]: As the file contained details regarding the constraint of the analysis, i don't think the file is necessary to include in the test model. 
Although the found violation in the file can be used later to test the full analysis

File: tests/org.palladiosimulator.dataflow.confidentiality.analysis.tests/src/org/palladiosimulator/dataflow/confidentiality/analysis/AnalysisUtils.java [hunk 3]
```diff
@@ -26,9 +36,16 @@ public static void assertSequenceElement(ActionSequence sequence, int index, Cla
         assertEquals(expectedType, actualType, createProblemMessage(index, expectedType, actualType));
     }
 
+    /**
+     * <em>Assert</em> that {@code sequence} and {@code expectedType} are of the same class at every index.
```
[the remark this answers is not in the document]
Comment (Entenwilli) [reply tests/org.palladiosimulator.dataflow.confidentiality.analysis.tests/src/org/palladiosimulator/dataflow/confidentiality/analysis/AnalysisUtils.java]: I have updated the documentation in an commit today to reflect that the order of sequence elements are taken into account here

File: tests/org.palladiosimulator.dataflow.confidentiality.analysis.tests/src/org/palladiosimulator/dataflow/confidentiality/analysis/DataFlowConfidentialityAnalysisTest.java [hunk 4]
```diff
@@ -2,58 +2,99 @@
...
                 TEST_MODEL_PROJECT_NAME, Activator.class, usageModelPath, allocationPath);
 
         analysis.initalizeAnalysis();
+	}
+
+	/**
+	 * Tests the standalone analysis by executing the ActionSequenceFinder. Fails if no sequences can be found
+	 */
+    @Test
+    public void testStandaloneAnalysis() {
```
[the remark this answers is not in the document]
Comment (Entenwilli) [reply tests/org.palladiosimulator.dataflow.confidentiality.analysis.tests/src/org/palladiosimulator/dataflow/confidentiality/analysis/DataFlowConfidentialityAnalysisTest.java:43]: I think it would be simpler and better organised to group the test models by functionality, because that would allow better debugging when a test fails, as functionality is grouped together. Some tests for edge cases the Label Propagation encounters would also be better organised in an grouping by functionality as some edge cases only occur in some models.

In return, the setup method for the analysis would not work anymore as it initializes one model, and testing by functionality in each class would require every model to be initialised.

[the remark this answers is not in the document]
Comment (sebinside) [reply tests/org.palladiosimulator.dataflow.confidentiality.analysis.tests/src/org/palladiosimulator/dataflow/confidentiality/analysis/DataFlowConfidentialityAnalysisTest.java:43]: I agree, lets's group the tests by functionality then.

[the remark this answers is not in the document]
Comment (Entenwilli) [reply tests/org.palladiosimulator.dataflow.confidentiality.analysis.tests/src/org/palladiosimulator/dataflow/confidentiality/analysis/DataFlowConfidentialityAnalysisTest.java]: If you take a look in the recent commits, i have implemented the test for the SequenceFinder in a test class. Currently, i am using a separate method to test each model, but parameterized testing would be possible. 

In implementing that for the `testCount` method however, i have realized that Eclipse does not run parameterized tests and does not show the test results. Should we continue using Parameterized Tests or change them to separate methods?

File: tests/org.palladiosimulator.dataflow.confidentiality.analysis.tests/src/org/palladiosimulator/dataflow/confidentiality/analysis/DataFlowConfidentialityAnalysisTest.java (hunk 1 above)
Comment (Entenwilli) [reply tests/org.palladiosimulator.dataflow.confidentiality.analysis.tests/src/org/palladiosimulator/dataflow/confidentiality/analysis/DataFlowConfidentialityAnalysisTest.java]: I would suggest removing them for now, but creating separate test cases in the future. For now i would like to get this PR merged, so i can implement the tests for the label propagation in the other pull request

File: tests/org.palladiosimulator.dataflow.confidentiality.analysis.tests/src/org/palladiosimulator/dataflow/confidentiality/analysis/DataflowSequenceFinderTest.java [hunk 5]
```diff
@@ -0,0 +1,155 @@
...
+
+import org.junit.jupiter.api.DisplayName;
+import org.junit.jupiter.params.ParameterizedTest;
+import org.junit.jupiter.params.provider.Arguments;
+import org.junit.jupiter.params.provider.MethodSource;
+import org.palladiosimulator.dataflow.confidentiality.analysis.sequence.entity.pcm.CallingSEFFActionSequenceElement;
+import org.palladiosimulator.dataflow.confidentiality.analysis.sequence.entity.pcm.CallingUserActionSequenceElement;
+import org.palladiosimulator.dataflow.confidentiality.analysis.sequence.entity.pcm.SEFFActionSequenceElement;
+
+public class DataflowSequenceFinderTest extends AnalysisFeatureTest {
```
Comment (sebinside) [inline tests/org.palladiosimulator.dataflow.confidentiality.analysis.tests/src/org/palladiosimulator/dataflow/confidentiality/analysis/DataflowSequenceFinderTest.java:21]: More precise would be `ActionSequenceFinderTest` as class name

File: tests/org.palladiosimulator.dataflow.confidentiality.analysis.tests/src/org/palladiosimulator/dataflow/confidentiality/analysis/DataflowSequenceFinderTest.java [hunk 6]
```diff
@@ -0,0 +1,155 @@
...
+
+    /**
+     * Provides a list of arguments to the {@code testPath} Test. The list contains the arguments to
+     * the test method
+     * 
+     * @return List of arguments to the test method
+     */
+    private static Stream<Arguments> testPathProvider() {
+        return Stream.of(
+                Arguments.of(onlineShopAnalysis,
```
Comment (sebinside) [inline tests/org.palladiosimulator.dataflow.confidentiality.analysis.tests/src/org/palladiosimulator/dataflow/confidentiality/analysis/DataflowSequenceFinderTest.java:72]: This is a hard to read and to maintain. I would propose to save the lists of element classes in properly named constants and then only referring to them here.

File: tests/org.palladiosimulator.dataflow.confidentiality.analysis.tests/src/org/palladiosimulator/dataflow/confidentiality/analysis/AnalysisFeatureTest.java [hunk 7]
```diff
@@ -0,0 +1,53 @@
...
+
+import static org.palladiosimulator.dataflow.confidentiality.analysis.AnalysisUtils.TEST_MODEL_PROJECT_NAME;
+
+import java.nio.file.Paths;
+
+import org.junit.jupiter.api.BeforeAll;
+import org.palladiosimulator.dataflow.confidentiality.analysis.testmodels.Activator;
+
+public class AnalysisFeatureTest {
+    protected static StandalonePCMDataFlowConfidentialtyAnalysis onlineShopAnalysis;
```
Comment (sebinside) [inline tests/org.palladiosimulator.dataflow.confidentiality.analysis.tests/src/org/palladiosimulator/dataflow/confidentiality/analysis/AnalysisFeatureTest.java:11]: Have these to be static?

**Review by sebinside — CHANGES_REQUESTED** [review]
Very cool approach with the parameterized test cases. I only have some minor and final remarks, addressing these should be finished quickly and then I will directly merge this PR. Thank you!

File: tests/org.palladiosimulator.dataflow.confidentiality.analysis.tests/src/org/palladiosimulator/dataflow/confidentiality/analysis/AnalysisUtils.java [hunk 8]
```diff
@@ -26,14 +44,27 @@ public static void assertSequenceElement(ActionSequence sequence, int index, Cla
...
+     * <p>
+     * If {@code sequence} is {@code null} or the sequences are of different length, they are
+     * considered unequal
+     * 
+     * @param sequence
+     *            ActionSequence to be inspected
+     * @param expectedType
+     *            Expected types of the given ActionSequence at all indexes
+     */
+    public static void assertSequenceElements(ActionSequence sequence, List<Class<?>> expectedElementTypes) {
```
[the remark this answers is not in the document]
Comment (Entenwilli) [reply tests/org.palladiosimulator.dataflow.confidentiality.analysis.tests/src/org/palladiosimulator/dataflow/confidentiality/analysis/AnalysisUtils.java:59]: As we use a list of classes as an input to the method now, it would make sense to do that. I`ve done that in one of the new commits

File: tests/org.palladiosimulator.dataflow.confidentiality.analysis.tests/src/org/palladiosimulator/dataflow/confidentiality/analysis/DataflowSequenceFinderTest.java (hunk 6 above)
Comment (Entenwilli) [reply tests/org.palladiosimulator.dataflow.confidentiality.analysis.tests/src/org/palladiosimulator/dataflow/confidentiality/analysis/DataflowSequenceFinderTest.java]: I agree. I`ve put the Paths for the test models into a separate class and changed the formatting to make it more clear, what data the Lists provide. Furthermore i`ve updated the documentation for the argument provider accordingly

File: tests/org.palladiosimulator.dataflow.confidentiality.analysis.tests/src/org/palladiosimulator/dataflow/confidentiality/analysis/AnalysisFeatureTest.java (hunk 7 above)
Comment (Entenwilli) [reply tests/org.palladiosimulator.dataflow.confidentiality.analysis.tests/src/org/palladiosimulator/dataflow/confidentiality/analysis/AnalysisFeatureTest.java]: With the `@TestInstance(Lifecycle.PER_CLASS)` in the Base Test, i got the test to work without the static attributes on the example analysis

File: tests/org.palladiosimulator.dataflow.confidentiality.analysis.tests/src/org/palladiosimulator/dataflow/confidentiality/analysis/AnalysisFeatureTest.java [hunk 9]
```diff
@@ -0,0 +1,53 @@
+package org.palladiosimulator.dataflow.confidentiality.analysis;
+
+import static org.palladiosimulator.dataflow.confidentiality.analysis.AnalysisUtils.TEST_MODEL_PROJECT_NAME;
+
+import java.nio.file.Paths;
+
+import org.junit.jupiter.api.BeforeAll;
+import org.palladiosimulator.dataflow.confidentiality.analysis.testmodels.Activator;
+
+public class AnalysisFeatureTest {
```
[the remark this answers is not in the document]
Comment (Entenwilli) [reply tests/org.palladiosimulator.dataflow.confidentiality.analysis.tests/src/org/palladiosimulator/dataflow/confidentiality/analysis/AnalysisFeatureTest.java]: I've changed the class name accordingly as that is more clear, on what the class is supposed to provide

File: tests/org.palladiosimulator.dataflow.confidentiality.analysis.tests/src/org/palladiosimulator/dataflow/confidentiality/analysis/DataflowSequenceFinderTest.java (hunk 5 above)
Comment (Entenwilli) [reply tests/org.palladiosimulator.dataflow.confidentiality.analysis.tests/src/org/palladiosimulator/dataflow/confidentiality/analysis/DataflowSequenceFinderTest.java]: I've changed the class name accordingly as `ActionSequenceFinderTest` is more clear, on what the class is supposed to test

**Review by sebinside — APPROVED** [review]
